sw/host/opentitansession/src/main.rs

```diff
@@ -0,0 +1,263 @@
+#[structopt(
+    name = "opentitansession",
+    about = "A tool for interacting with OpenTitan chips."
+)]
+struct Opts {
+    #[structopt(
+        long,
+        default_value = "config",
+        help = "Filename of a default flagsfile.  Relative to $XDG_CONFIG_HOME/opentitantool."
+    )]
+    rcfile: PathBuf,
+
+    #[structopt(long, default_value = "off")]
+    logging: LevelFilter,
+
+    #[structopt(flatten)]
+    backend_opts: backend::BackendOpts,
+
+    #[structopt(long)]
+    stop: bool,
+
+    #[structopt(long)]
+    listen_port: Option<u16>,
+
+    #[structopt(long, help = "Start session, staying in foreground (do not daemonize)")]
+    debug: bool,
+}
+
+// Given some existing option configuration, maybe re-evaluate command
+// line options by reading an `rcfile`.
+fn parse_command_line(opts: Opts, mut args: ArgsOs) -> Result<Opts> {
+    // Initialize the logger if the user requested the non-defualt option.
+    let logging = opts.logging;
+    if logging != LevelFilter::Off {
+        env_logger::Builder::from_default_env()
+            .filter(None, opts.logging)
+            .init();
+    }
+    if opts.rcfile.as_os_str().is_empty() {
+        // No rcfile to parse.
+        return Ok(opts);
+    }
+
+    // Construct the rcfile path based on the user's config directory
+    // (ie: $HOME/.config/opentitantool/<filename>).
+    let rcfile = if let Some(base) = ProjectDirs::from("org", "opentitan", "opentitantool") {
+        base.config_dir().join(&opts.rcfile)
+    } else {
+        opts.rcfile
+    };
+
+    // argument[0] is the executable name.
+    let mut arguments = vec![args.next().unwrap()];
+
+    // Read in the rcfile and extend the argument list.
+    match std::fs::read_to_string(&rcfile) {
+        Ok(content) => {
+            for line in content.split('\n') {
+                // Strip basic comments as shellwords won't handle comments.
+                let (line, _) = line.split_once('#').unwrap_or((line, ""));
+                arguments.extend(shellwords::split(line)?.iter().map(OsString::from));
+            }
+            Ok(())
+        }
+        Err(e) if e.kind() == ErrorKind::NotFound => {
+            log::warn!("Could not read {:?}. Ignoring.", rcfile);
+            Ok(())
+        }
+        Err(e) => Err(anyhow::Error::new(e).context(format!("Reading file {:?}", rcfile))),
+    }?;
+
+    // Extend the argument list with all remaining command line arguments.
+    arguments.extend(args.into_iter());
+    let opts = Opts::from_iter(&arguments);
+    if opts.logging != logging {
+        // Try re-initializing the logger.  Ignore errors.
+        let _ = env_logger::Builder::from_default_env()
+            .filter(None, opts.logging)
+            .try_init();
+    }
+    Ok(opts)
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct SessionStartResult {
+    port: u16,
+}
+
+/// Fork a child process, and in the child process instantiate Transport based on command line
```

Comment:
I would really prefer to avoid fork. Fork is basically impossible to use safely. I would much prefer if you made the binary re-exec itself with different command line arguments.

This will also let you do such things as null out fds 012 as you do below, and other conveniences.

Reply:
I considered that (and also considered using a library such as https://docs.rs/daemonize/latest/daemonize/ ).
One issue with traditional subprocess exec is that it is not straightforward how to "reconstruct" the parts of the command line arguments that go into the backend_opts, which the subprocess is going to need.

Manual subprocess would probably still need "nix" call to setsid() to avoid the subprocess being killed, if the user closes the terminal window used to start it.  This could be nicer if using a proper daemonize library, but the one I found (linked above) did not seem to be able to handle the port number being part of the pid file, which we need.

I do realize that you have to be careful with fork(), an early iteration for instance created the non-trivial Transport object before the fork(), and I had to consider forget()-ing that object in the parent process, to prevent drop() from e.g. sending shutdown commands to the USB device, preventing the child process from using its clone of the Transport object.  I think I have arrived at a code structure here, where there is very little sharing of objects between the parent and child, basically just the parsed representation of the command line options, backend_opts, is "cloned", and all the non-trivial objects are created by the child only after fork()-ing.  I foresee little change necessary in that division of responsibility going forward, so I was hoping that this solution could be acceptable.

Reply:
I'm not sure I follow... why is it hard to reconstruct them? I would have expected you would add either a subcommand or something that you would call when self-execing, which the child could use to extract the command line arguments passed from the parent.

I do not thing using fork() is a good idea, since it does not work with multi-threaded programs and will make this code unmaintainable in the long run, since random parts will essentially need to be async-signal-safe. Please consider restructuring this code to self-exec as I suggested; given the existing structure it should be doable.